src/org/jitsi/impl/neomedia/rtp/translator/RTCPFeedbackMessageSender.java

```diff
@@ -102,87 +113,291 @@ private long getSenderSSRC()
 
     /**
      * Sends a Full Intra Request (FIR) Command to a media sender/source with a
-     * specific synchronization source identifier (SSRC).
-     * <p>
-     * <b>Warning</b>: Due to (current) limitations in
-     * <tt>RTPTranslatorImpl</tt> and/or <tt>StreamRTPManager</tt>, a FIR
-     * command with the specified <tt>mediaSenderSSRC</tt> is sent to all
-     * <tt>MediaStream</tt>s.
-     * </p>
+     * specific synchronization source identifier (SSRC) through a specific
+     * <tt>MediaStream</tt>.
      *
      * @param mediaSenderSSRC the SSRC of the media sender/source
      * @return <tt>true</tt> if a FIR command was sent; otherwise,
      * <tt>false</tt>
      */
     public boolean sendFIR(int mediaSenderSSRC)
     {
-        boolean sentFIR = false;
-        /*
-         * XXX Currently, the method effectively broadcasts a FIR i.e. sends it
-         * to all streams connected to rtpTranslator because MediaStream's
-         * method getRemoteSourceIds returns an empty list (possibly due to
-         * RED).
-         */
-        for (StreamRTPManager streamRTPManager
-                : rtpTranslator.getStreamRTPManagers())
+        boolean registerRecurringProcessible = false;
+        synchronized (kfRequesters)
```

Comment:
The normal non-synchronised approach using ConcurrentHashMap would be:

    KeyframeRequester kfRequester = kfRequesters.get(mediaSenderSSRC);
    if (kfRequester == null)
    {
        // No existing requester; create a new one
        kfRequester = new KeyframeRequester(mediaSenderSSRC);
        KeyframeRequester existing = kfRequesters.putIfAbsent(mediaSenderSSRC, kfRequester);
        if (existing != null)
        {
            // Another thread created one at the same time, and beat this one
            kfRequester = existing;
        }
    }

This way you avoid locking, but you don't continue to create new instances long after one becomes available.